crypto/zkp/primitives/commitments/damgard-fujisaki_square.go

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2017 XLAB d.o.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package commitmentzkp
+
+import (
+	"math/big"
+
+	"fmt"
+
+	"github.com/xlab-si/emmy/crypto/commitments"
+)
+
+type DFCommitmentSquareProver struct {
+	equalityProver *DFCommitmentEqualityProver
+}
+
+func NewDFCommitmentSquareProver(committer *commitments.DamgardFujisakiCommitter,
+	x *big.Int, challengeSpaceSize int) (*DFCommitmentSquareProver, *big.Int, error) {
+
+	// Input committer contains c = g^(x^2) * h^r (mod n).
+	// We now create two committers - committer1 will contain c1 = g^x * h^r1 (mod n),
+	// committer2 will contain the same c as committer, but using a different base c = c1^x * h^r2.
+	// Note that c = c1^x * h^r2 = g^(x^2) * h^(r1*x) * h^r2, so we choose r2 = r - r1*x.
+	// DFCommitmentSquareProver proves that committer1 and committer2 hide the same value (x) -
```

Comment:
Can you move these comments to more appropriate place? For instance I think L39-L40 should be put above  `DFCommitmentSquareProver`.

Reply:
The comment describes the internals of the proof so the place is appropriate (it is not something that the user of a function needs to know, but I prefer to have such explanations available in the code). L39 and L40 depend on the lines before. But I added a comment to the prover struct.